### ISSUE TITLE
GuardedBy false positive with Runnable passed to a user method known to invoke synchronously

### ISSUE THREAD

**Issue by EamonTracey** [google/error-prone#5938]
The `@GuardedBy` annotation has a false positive when we pass a lambda or method reference that accesses guarded items, even if we know it is invoked immediately synchronously.

Consider the following case:
```java
class Transaction {
  @GuardedBy("this")
  int x;

  public synchronized void handle() {
    doSomething(() -> {
      x++;  // compilation error, requires extraneous synchronized block to satisfy the checker
    });
  }

  private void doSomething(Runnable r) {
    r.run();
  }
}
```

The above code fails to compile without adding a `synchronized` block inside the lambda. This grows cumbersome in codebases where this pattern is common, for example, a `safeRun` method that wraps a `Runnable` with error handling and logging.

I suggest we add a parameter annotation `@RunsImmediately` to indicate the call is safe without the extraneous synchronization block. There already exists a [whitelist for common built-in Java functions](https://github.com/google/error-prone/blob/be8896aee16f3de5f95d188f3ba1045e36e872ac/core/src/main/java/com/google/errorprone/bugpatterns/threadsafety/HeldLockAnalyzer.java#L68-L89), so this would allow people to whitelist their own methods.

**Comment by cushon** [comment]
Instead of an annotation, we could add a flag to the check to make that list of well known APIs configurable.

For the error prone annotations artifact we try to guarantee there will never be breaking changes to the annotations, and there’s a higher bar to adding annotations compared to flags.

**Comment by EamonTracey** [comment]
Makes sense, will update the PR to introduce a new flag rather than annotation
